### DISCUSSION TITLE
Multiroom / Grouped Speakers synchronized audio

### DISCUSSION THREAD

**Discussion by foxy82** [music-assistant/support#385] [Feature requests and ideas]
Hey - thanks so much for the work on this and I'd be happy to help out.

For me the main thing that is missing in the open source audio solutions is a true multiroom experience where any stream can be played to any/many speakers. 

There are a few options that cover some of this but nothing seems to get  it 100%.
1) Snapcast - great for distribution but requires other software for setting up/controlling sources - e.g. TuneIn, Spotify. No meta data at the clients
2) Shairport-sync / owntone - No easy way to do multiple streams / multiple speaker groups.

Do you have a view on how you are looking to do this? I had a look in the issues but couldn't see anything related to this. I know you mentioned it on the launch video. 

I'd be happy to help out.
Thanks

**Comment by marcelveldt** [comment]
Playing the same audio to multiple speakers: already supported. Just create a group helper in HA and start playback to that group. Powered clients will start outputting the audio.

The audio however is not yet synchonized. I had planned to have that in the first release but there was not enough time to finish it.
This sync feature is going to work with players that provide accurate playback timestamps, e.g. slimproto (squeezebox), airplay, snapcast and maybe others too.

This is something I will start to implement next weeks and provide a beta version to test.

**Reply by erkr** [reply]
Related to feature #410

**Reply by richardwonka** [reply]
I frequently (many times a day) re-group my speakers, depending on where i am in the flat, so having static groups isn't a good solution here, not even a good workaround.

would it not be easy to have buttons that call HA's built-in services `media_player.join` and `media_player.unjoin`  next to each player?

**Reply by marcelveldt** [reply]
I think you are misunderstanding the feature. You will have a group like "whole house audio" to which you can start playing.
Music will only be played at speakers that are "powered on" in that group, meaning that the joining of players is automatically done for you.

If you don't want to use the easy groups and do the hard work yourself by manually joining/unjoining that will be provided too.

**Reply by sorryusernameisalreadytaken** [reply]
goede dag @marcelveldt :)

Is there any news regarding the synchronisation problem?

I love running slimproto with music-assistant on one esp but would love to have it running syncronously on several.

**Reply by OzGav** [reply]
Have you tried using a universal group?

**Reply by ryanwwest** [reply]
Also interested in the sync problem's progress. Last update here was Feb 11 (https://github.com/orgs/music-assistant/discussions/385#discussioncomment-4890774) saying that some effort was made but it might be difficult to solve—would be interesting to know if this will be pursued more in the near future or if it's not worth it.

**Reply by OzGav** [reply]
No additional progress has been made as no new player types have been added. What players are you trying to sync?

**Reply by ryanwwest** [reply]
Thanks. Not sure if the player types question was directed to me, but I'm just generally curious if this could be a suitable replacement compared to something like the Soundavo WS66i 6-source 12-channel matrix amp which (I think) solves sync issues. I'd prefer this if so, but maybe it's not ideal for now if different zones/stereo speaker sets are very close to one another.

**Reply by OzGav** [reply]
My point about new player types is that some can have perfect sync and some can’t depending on the capabilities of the player (accurate time stamps). You can review the FAQs on GitHub for each player type to see what is supported. Not sure about the Soundavo you will need to do a feature comparison of each. Ask any more specific questions you need to understand what MA can do (noting if it is still in beta and more features will be added in time - and some bugs still exist)

**Reply by foxy82** [reply]
How about incorporating snapcast?

**Reply by sorryusernameisalreadytaken** [reply]
Just so I understand this correctly. There is just no way for me to run [squeezelite-esp32](https://github.com/sle118/squeezelite-esp32) in multiple rooms and run it synchronously with the help of [slimproto](https://www.home-assistant.io/integrations/slimproto/) and music-assistant?

I apologise if I am misunderstanding this.

**Reply by OzGav** [reply]
Slimproto does support perfect sync so you can do what you are asking. You just need to use MA only

**Reply by marcelveldt** [reply]
@sorryusernameisalreadytaken no, you only need music assistant for this or ofcourse the original Logitech Media Server from which slimproto originated from. You do NOT need the slimproto integretion in HA, that is only for very basic control of the devices.

**Reply by marcelveldt** [reply]
@foxy82  snapcast is somewhere on the wishlist, that is a very good fit to integrate into MA

**Comment by achusuresh2** [comment]
When this feature is implemented, would it also be possible to implement timing correction as well for devices that are playing out of sync. Currently google's default timing correction is liming to 150ms, it would be great if music assistant can provide more timing correction options for google cast devices.

**Reply by marcelveldt** [reply]
sync is platform dependent. So Sonos speakers play in sync, cast speakers play in sync, airplay, snapcast, slimproto etc.
You can't sync e.g. a cast speaker to an airplay one. Also the timing correction is builtin cast so nothing we can do about that.

I actually toyed for some time trying to get those platforms in sync but they provide such an inaccurate timestamp information that is is useless. Airplay on the other hand does provide accurate timestamps so that can be synced to slimproto based players which also have accurate timestamps. Maybe snapcast too but did not yet look into that.

At some point I might provide some "DIY sync" options where you can toy with start delay and drift correction settings but I doubt that will bring a good experience if a player does not provide any accurate information of what pcm frames it did actually play at what time.

**Comment by ddppddpp** [comment]
I'm running an owntone setup where I can use multi-room/multi-speaker with synchronization via owntone or address each speaker individually for multiple streams.

**Comment by that1guy** [comment]
I see #410 was moved to this discussion.  When I read the conversations I get a bit confused so I wanted to make sure I’m understanding the feature correctly.

Are we saying:

“As a Mass user I want to be able to group speakers together without having to pre-define a helper group”?

Ideally, this is the feature I’m hunting for.  If Mass can do this then I don’t need to use Sonos anymore.

**Reply by marcelveldt** [reply]
Yes, MA will get its own universal grouping type where you can dynamically add/remove speakers from all ecosystems with some sync strategies. The helper group support will stay but it doesn't support member management from service calls.

**Reply by NWSS** [reply]
I'm looking forward to this feature !!! :-) Dit lijkt me een zware dobber om te maken! Ik ga sowieso een koffie kopen voor je ;-)

**Reply by richardwonka** [reply]
That's exactly what I would need, too. dynamic grouping is a must in my use case

**Comment by that1guy** [comment]
Thanks for confirming @marcelveldt.  Very exciting news!

**Comment by sorryusernameisalreadytaken** [comment]
I hope to see that feature soon. Thanks for your great project here :)

**Comment by DesignsWithDavid** [comment]
I am willing to help develop the feature, but I am unclear what has been acomplished and what remains.  @marcelveldt -- Are you able to provide a status update on this?  Thank you for all your work!  Please let me know how I can help!

**Reply by marcelveldt** [reply]
I am currently working on a large refactor to improve a lot of stuff and implement a few features that would otherwise been hard to implement. I'd like to call this the v2 release of Music Assistant. Once that is released I will also write down some proper docs for development as I have improved so many models and interfaces that makes it way easier to contribute to the project.

As for multi room sync. I have implemented multi room sync for squeezebox (slimproto) players, sonos and airplay. This means that speakers of any of those platforms can play in perfect sync. slimproto and airplay can even be synced to eachother.

Next to the sync feature will be the grouping feature where you can create groups of speakers. You can play music to such a group and music will start playing to all players in the group (or depending on a setting, only the powered ones). If there are syncable players in that group those will play in perfect sync, others won't.

Trying to sync e.g. a Chromecast player to some random DLNA or Airplay player is not possible and will never work 100%. They will probably start playing at more or less the same time but then will get slowly drift apart.

**Reply by DesignsWithDavid** [reply]
Wow!  That is quite the effort!

I look forward to seeing and contributing to V2.  Tha ns again for all your hard work

**Reply by jmvaz** [reply]
And sync google homes together will it work?

**Reply by OzGav** [reply]
Based on what Marcel said here I believe the answer is yes https://github.com/music-assistant/hass-music-assistant/discussions/385#discussioncomment-4890774

**Reply by marcelveldt** [reply]
Correct, that will work the same as now. Just create a speaker group in the Google Home app.
As far as I know there is not (yet) an API to programmatically create Cast speakers groups so you will have to do that yourself manually

**Reply by dexdeadly** [reply]
You can also add google branded devices in a group and they will play no problem.  However a google cast device such as a vizio sound bar like I have and tested, will not be in sync.  It seems to start behind about 3 seconds behind.  You can create the group in home assistant or google home app, either way its the same result.

**Reply by marcelveldt** [reply]
That sounds like an awful bug in that device's implementation of the cast protocol. BTW: You can't create a cast group from within Home Assistant. You need the Google Home app top create a real sync group of cast speakers. 

You'll have to create support ticket with the device manufacturer for that. I have tested a lof of devices (most of them with a cast and/or airplay implementation onboard) and its the first time I've head of this. Maybe you have a defective one ?

**Comment by marcelveldt** [comment]
This feature request should actually be closed because this feature is implemented in Music Assistant (beta) 2.0 where you can create a Universal Group and that will automatically sync speakers from the same ecosystem, meaning it can do perfect sync for sonos devices, slimproto based devices and airplay. It also supports Google cast groups natively so plenty of options. 

For sync support of speakers from different ecosystems I can be short: Not going to work ever. I might one day release my optimistic approach for that but without accurate timestamps it is impossible and if you ask me you just shouldn't want it. Just make sure that you group audio outputs of the same protocol within hearing distance, that will give the best experience.

To add support for other platforms that provide syncing features, like snapcast and linkplay, we have some additional feature requests to add player support for those platforms.

**Reply by simowilso** [reply]
I'm new to Music Assistant and got excited when I saw this thread. Then I saw " Not going to work ever". What a pity - that would have been a killer feature, but I guess may have ended up in a possible legal issue too ;)
